### ISSUE TITLE
conch client completely broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10884]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#1292|
|Type|defect|
|Created|2005-10-29 02:28:27Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1292 1292
type__defect defect
reporter__glyph glyph
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__backport_soc2007 backport soc2007
time__1130552907000000 1130552907000000
changetime__1226804104000000 1226804104000000
version__ 
owner__ 
cc__glyph cc__exarkun cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Depending on the state of ~/.ssh/known_hosts, this is different.  However, it
isn't a format bug, because the traceback still happens (and I can't connect)
regardless of whether the file exists but is empty, exists but has one empty
line, exists and contains openssh known hosts, or doesn't exist at all.

Here's a sample traceback:


glyph@legion:~% Projects/Twisted/trunk/bin/conch/conch --known-hosts=empty
divmod.com
The authenticity of host 'divmod.com' can't be established.
RSA key fingerprint is 57:a1:c2:a1:07:a0:2b:f4:ce:b5:e5:b7:ae:cc:e1:99.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'divmod.com' (RSA) to the list of known hosts.
Traceback (most recent call last):
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/log.py", line 58, in
callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/log.py", line 43, in
callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 59,
in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 37,
in callWithContext
    return func(*args,**kw)
--- &lt;exception caught here> ---
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/selectreactor.py",
line 139, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/tcp.py", line 347,
in doRead
    return self.protocol.dataReceived(data)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line
187, in dataReceived
    f(packet[1:])
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line
534, in ssh_KEX_DH_GEX_REPLY
    d = self.verifyHostKey(pubKey, fingerprint)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/client/direct.py", line
77, in verifyHostKey
    fingerprint)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/client/default.py",
line 48, in verifyHostKey
    if known_hosts.read(1) != '\n':
exceptions.IOError: [Errno 9] Bad file descriptor
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
partially fixed by r14953
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|z3p commented|
|-|-|

Tagging soc2007

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p commented|
|-|-|

I'm not sure what to make of this.  Trying it today with your command line functions as expected.  Does this still happen for you?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

http://twistedmatrix.com/trac/changeset/14953 probably prevented the exception from showing up, but it included no unit tests, and it might not be the appropriate fix for the problem.  This ticket can probably be resolved by adding the unit tests and making whatever further implementation changes (if there are any) that process suggests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

[#10942](https://github.com/twisted/twisted/issues/10942) probably fixes this as well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `closed`|
|-|-|

(In [25365]) Fixes for the conch client's known_hosts parsing and host key verification.

Fixes [#10942](https://github.com/twisted/twisted/issues/10942)
Fixes [#10890](https://github.com/twisted/twisted/issues/10890)
Fixes [#3494](https://github.com/twisted/twisted/issues/3494)
Fixes [#3496](https://github.com/twisted/twisted/issues/3496)
Fixes [#10884](https://github.com/twisted/twisted/issues/10884)
Fixes [#3499](https://github.com/twisted/twisted/issues/3499)

The main thrust of this change is that it adds a new module to conch, `twisted.conch.client.knownhosts`, which provides a structured representation of OpenSSH's `known_hosts` file.  This is a big step in the direction of modularizing conch's support for verifying host keys and storing the results of that verification, although the internal connection APIs are mostly unchanged at this point, they could now easily be adjusted to speak to an abstract interface and still manipulate the user's actual known_hosts entries rather than completely overriding them.

The individual bugs which this change fixes are too numerous to bother describing twice; have a look at the tickets themselves more information.  Suffice it to say that the conch client is now a lot more reliable (although it still has a long way to go).

In addition to making the conch client more stable, the new API allows for some programmatic manipulation of known_hosts files, which might be independently useful.

Additionally, although the review notes some ways in which the tests can still be improved, the tests for this module can be cited as a _much_ better example for conch contributors as to how unit tests within conch should look.  While these tests are covering almost exclusively new code, they are covering functionality which has existed (and been broken) for quite some time.

Author: glyph

Reviewers: exarkun, jml

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
